### PR TITLE
Add qvector, qarray, and qview

### PR DESCRIPTION
**Pull request by amccaskey**

### CODE REVIEW

File: runtime/cudaq/builder/kernel_builder.cpp [hunk 1]
```diff
@@ -97,6 +97,11 @@ KernelBuilderType mapArgToType(cudaq::qreg<> &e) {
       [](MLIRContext *ctx) mutable { return quake::VeqType::getUnsized(ctx); });
 }
 
+KernelBuilderType mapArgToType(cudaq::qvector<> &e) {
+  return KernelBuilderType(
+      [](MLIRContext *ctx) mutable { return quake::VeqType::getUnsized(ctx); });
```
Comment (schweitzpgi) [inline runtime/cudaq/builder/kernel_builder.cpp:102]: Does this really need to be marked `mutable`? The lambda has no captures.

File: runtime/cudaq/builder/kernel_builder.cpp [hunk 2]
```diff
@@ -479,23 +484,23 @@ QuakeValue applyMeasure(ImplicitLocOpBuilder &builder, Value value,
   return QuakeValue(builder, ret);
 }
 
-QuakeValue mx(ImplicitLocOpBuilder &builder, QuakeValue &qubitOrQreg,
+QuakeValue mx(ImplicitLocOpBuilder &builder, QuakeValue &qubitOrQvec,
```
Comment (schweitzpgi) [inline runtime/cudaq/builder/kernel_builder.cpp:487]: nit: I'm ok with abbreviations and this makes sense to me. Some people like to see names typed out entirely everywhere. Which I think would mean some variation of `qvector` for these. Mentioning it should we document coding style guidelines for runtime, compiler, etc.

File: runtime/cudaq/qis/qarray.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+private:
+  /// @brief Reference to the held / owned array of qudits
+  std::array<value_type, N> qudits;
+
+public:
+  /// Nullary constructor
+  qarray() {}
+
+  /// @brief qarray cannot be copied
+  qarray(qarray const &) = delete;
```
Comment (schweitzpgi) [inline runtime/cudaq/qis/qarray.h:41]: nit: consider `const` on left side here (assuming that is more consistent with other code).

**Review by schweitzpgi — COMMENTED** [review]
still reviewing this one...

**Review by schweitzpgi — COMMENTED** [review]
still reviewing this one...

File: runtime/cudaq/qis/qarray.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  /// @brief Returns the `[count, size())` qudits as a non-owning qview
+  qview<Levels> back(std::size_t count) {
+    return std::span(qudits).subspan(size() - count, count);
+  }
+
+  /// @brief Returns the last qudit.
+  value_type &back() { return qudits.back(); }
+
+  /// @brief Returns the `[start, start+size)` qudits as a non-owning qview
+  qview<Levels> slice(std::size_t start, std::size_t size) {
```
Comment (schweitzpgi) [inline runtime/cudaq/qis/qarray.h:72]: Do we ever want to provide a strided slice for things like "every even numbered element between 10 and 19"?

File: runtime/cudaq/qis/qreg.h [hunk 5]
```diff
@@ -25,7 +25,10 @@ concept ValidQregSize = N > 0;
 /// template parameters.
 template <std::size_t N = dyn, std::size_t Levels = 2>
   requires(details::ValidQregSize<N>)
-class qreg {
+class
+    // [[deprecated(
+    // "qreg deprecated and will be removed. use qvector or qarray.")]]
```
Comment (schweitzpgi) [inline runtime/cudaq/qis/qreg.h:30]: We should probably have an action item to remove this deprecated feature.

File: runtime/cudaq/qis/qvector.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  /// meant to be used with kernel_builder<cudaq::qvector<>>
+  qvector() : qudits(1) {}
+  /// @endcond
+
+  /// @brief qvectors cannot be copied
+  qvector(qvector const &) = delete;
+
+  /// @brief qvectors cannot be moved
+  qvector(qvector &&) = delete;
+
```
Comment (schweitzpgi) [inline runtime/cudaq/qis/qvector.h:43]: Do you want to delete the assignment operator?

**Review by schweitzpgi — COMMENTED** [review]
LGTM. Thanks for cleaning this up, Alex.

File: runtime/cudaq/qis/qarray.h (hunk 4 above)
Comment (amccaskey) [reply runtime/cudaq/qis/qarray.h:72]: We probably do, I'll add an issue for it and we can add it in another PR.

File: runtime/cudaq/qis/qreg.h (hunk 5 above)
Comment (amccaskey) [reply runtime/cudaq/qis/qreg.h:30]: Will do

**Review by schweitzpgi — APPROVED** [review]
+1

**Comment by bettinaheim** [comment]
@amccaskey Where are the matching documentation changes?

**Pull request merged**